### DISCUSSION TITLE
Slow LazyVim - looking for assistance in profiling the cause(s)

### DISCUSSION THREAD

**Discussion by yanshay** [LazyVim/LazyVim#2196] [Q&A]
My LazyVim based Neovim is very slow, same projects I open with VSCode are so much more responsive, like night and day. And I refer to even simple operations such as scrolling up/down, moving cursor across rows with J/K. It sometimes gets to taking around 1/2 sec to move between rows.

I'm using mostly rust, but it is also similar in other projects.

So I tried using the profiler at a time when response was very slow, I profiled only up/down movements across rows for around 10-20 seconds I think, nothing more than that to make it simple. I don't know how exactly to profile further, so assistance would be welcomed:

Here is the profiling summary:
```
FUNCTIONS SORTED ON SELF TIME
count  total (s)   self (s)  function
  514              2.248688  <SNR>1_LoadFTPlugin()
  514              0.966529  <SNR>35_SynSet()
  514              0.406593  <SNR>2_LoadIndent()
  803   0.128163   0.115528  <SNR>24_Highlight_Matching_Pair()
    5              0.030480  nvim_treesitter#indent()
  805              0.012668  <SNR>24_Remove_Matches()
   24   0.558489   0.005033  <SNR>46_try_cmd()
   21   0.468996   0.001259  3()
   24   0.561168   0.001200  provider#clipboard#Call()
    3   0.090972   0.000220  2()

```

This seems strange, why is the LoadFTPlugin being called so many times and take so much time when just moving a cursor across rows?

When I looked at the LoadFTPLugin section it includes the below, which reference lua, but my project is rust, not lua.
```
FUNCTION  <SNR>1_LoadFTPlugin()
    Defined: /usr/local/Cellar/neovim/0.9.1/share/nvim/runtime/ftplugin.vim:14
Called 514 times
Total time:   2.248688
 Self time:   2.248688

count  total (s)   self (s)
  514              0.002936     if exists("b:undo_ftplugin")
                                  exe b:undo_ftplugin
                                  unlet! b:undo_ftplugin b:did_ftplugin
  514              0.000471     endif
                            
  514              0.002444     let s = expand("<amatch>")
  514              0.000939     if s != ""
  514              0.004364       if &cpo =~# "S" && exists("b:did_ftplugin")
                            	" In compatible mode options are reset to the global values, need to
                            	" set the local values also when a plugin was already used.
                            	unlet b:did_ftplugin
  514              0.000308       endif
                            
                                  " When there is a dot it is used to separate filetype names.  Thus for
                                  " "aaa.bbb" load "aaa" and then "bbb".
 1028              0.004306       for name in split(s, '\.')
                                    " Load Lua ftplugins after Vim ftplugins _per directory_
                                    " TODO(clason): use nvim__get_runtime when supports globs and modeline
  514              0.394759         exe printf('runtime! ftplugin/%s.vim ftplugin/%s.lua', name, name)
  514              1.253084         exe printf('runtime! ftplugin/%s_*.vim ftplugin/%s_*.lua', name, name)
  514              0.570897         exe printf('runtime! ftplugin/%s/*.vim ftplugin/%s/*.lua', name, name)
 1028              0.003932       endfor
  514              0.000397     endif
```

Synset which was seond on the list in the summary:
```
FUNCTION  <SNR>35_SynSet()
    Defined: /usr/local/Cellar/neovim/0.9.1/share/nvim/runtime/syntax/synload.vim:26
Called 514 times
Total time:   0.966529
 Self time:   0.966529

count  total (s)   self (s)
                              " clear syntax for :set syntax=OFF  and any syntax name that doesn't exist
  514              0.001776   syn clear
  514              0.001285   if exists("b:current_syntax")
                                unlet b:current_syntax
  514              0.000345   endif
                            
  514              0.001894   0verbose let s = expand("<amatch>")
  514              0.000773   if s == "ON"
                                " :set syntax=ON
                                if &filetype == ""
                                  echohl ErrorMsg
                                  echo "filetype unknown"
                                  echohl None
                                endif
                                let s = &filetype
  514              0.000671   elseif s == "OFF"
                                let s = ""
  514              0.000223   endif
                            
  514              0.000504   if s != ""
                                " Load the syntax file(s).  When there are several, separated by dots,
                                " load each in sequence.  Skip empty entries.
 1028              0.004047     for name in split(s, '\.')
  514              0.001147       if !empty(name)
  514              0.477256         exe "runtime! syntax/" . name . ".vim syntax/" . name . "/*.vim"
  514              0.462544         exe "runtime! syntax/" . name . ".lua syntax/" . name . "/*.lua"
  514              0.000861       endif
 1028              0.003016     endfor
  514              0.000291   endif
```

**Comment by abeldekat** [comment]
Hi @yanshay,

I can't comment on your profiling unfortunately.

It would be interesting to know if the problem is present in stock LazyVim starter. 

``` lua
1. `mkdir ~/.config/lazytest`
2. `git clone https://github.com/LazyVim/starter ~/.config/lazytest`
3. `NVIM_APPNAME=lazytest nvim` and wait for installation of plugins to finish
4. Restart Neovim using the same command as in step 3
```

If it isn't, the problem likely is in your config and/or in the extra plugins you use. You could manually disable some plugins. 
You could also try [LazyFlex](https://github.com/abeldekat/lazyflex.nvim), a plugin I created.

**Comment by alanscodelog** [comment]
I've also been running into something similar for a while. Up/Down movements lagging randomly. I've tried everything. Problem is not there in plain neovim, but is there with the lazy starter (but then the plain neovim doesn't have any lsp, etc configured). My profiling looks very similar to op, a long time spent in ftplugin which is the filetype plugin I think. The weird thing is if I set filetype off and syn off just in case. I get zero seconds spent on any of those functions, but I still get the lag. It's so weird.

**Comment by tkkcc** [comment]
LazyVim scrolling is lag for me too. The frame drop occurs on treesitter-context content change. After `TSContextDisable`, the lag seems gone. But When I tried the mininal config on `treesitter-context`'s issue page, it's smooth too, and with context working.

**Comment by KerekesDavid** [comment]
I think I have a very similar problem, but treesitter-context is not installed with my lazyvim. How would I go about profiling it?

**Comment by nagy135** [comment]
i also see lag when holding "j" ...sometimes it seems to lag and repaint entire buffer as if it was an old tv screen, line by line. The issue i have is that its really hard to pin-point where it comes from. This isnt an big issue, holding "j" is antipattern anyway, but i would really love to see some profiling on performance, some tool that would just say "in last 5 seconds, we spent 70% of the time rendering context breadcrumbs, 5% on highlighting, ..." or something like that.
